ably/rest_client.go

```diff
@@ -43,9 +43,67 @@ func query(fn func(string, interface{}) (*http.Response, error)) QueryFunc {
 	}
 }
 
-type RestClient struct {
-	Auth *Auth
+// RestChannels provides an API for managing collection of RestChannel. This is
+// safe for concurrent use.
+type RestChannels struct {
+	cache  map[string]*RestChannel
+	mu     sync.RWMutex
+	client *RestClient
+}
 
+// Range iterates over the channels calling fn on every iteration. If fn returns
+// false then the iteration is stopped.
+func (c *RestChannels) Range(fn func(name string, channel *RestChannel) bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	for k, v := range c.cache {
```

Comment:
If anybody calling `Range` tries to modify `c` in `fn` (via either `Release` or "indirectly" via `Get`), this would lead to a deadlock. One way to handle this would be to:

* create a temporary map with size `len(c.cache)`
* `RLock`
* "clone" `c.cache` into the temp map with a `range`
* `RUnlock`
* `range` over the temp map and call `fn` at this point

This way concurrent modification from inside `fn` won't cause deadlocks. You should probably also add a test where you `Release` a channel/`Get` a channel that's not in `cache` from inside `fn`